---
title: "Figure 3: MIC-Drop-seq Analysis - meox1 Phenotype Validation"
author: "Clay Carey"
date: "2024-09-23"
output: html_document
---

```{r setup}
# Set working directory for analysis
knitr::opts_knit$set(root.dir = "/Users/claytoncarey/Documents/Seurat/micdrop_seq/micdrop/Figure_code/Code_for_submission/MIC-Drop-seq")
```

## Load Required Libraries
```{r load-libraries, message=FALSE, warning=FALSE}
# Data manipulation and visualization
library(Seurat)
library(dplyr)
library(ggplot2)
library(tidyverse)

# Single-cell specific
library(scCustomize)

# Visualization
library(viridis)
library(cowplot)
library(ggpubr)

# Statistics
library(rstatix)
library(DescTools)
```

## Load Data
```{r load-data}
# Load processed MIC-Drop-seq Seurat object
micdrop <- readRDS(file = "input_data/micdrop_2-6-25.rds")
```

# ============================================================================
# SECTION 1: DOTPLOT - meox1 TARGET GENE EXPRESSION IN SOMITIC MESODERM
# ============================================================================

```{r figure3A-dotplot}
# Set default assay and identities
DefaultAssay(micdrop) <- "RNA"
Idents(micdrop) <- "curated_cell_type2"

# Define somitic mesoderm cell types of interest
som_ct <- c("Sclerotome", 
            "Sclerotome - Chondrogenic", 
            "Myotome") 

# Define meox1 target genes to visualize
meox_genes <- c("tcf15", "cthrc1a", "nid1a", "vwde", 
                "thbs4b", "cilp", "comp")

# Create binary classification: meox1 mutants vs all others
micdrop@meta.data <- micdrop@meta.data %>%
  mutate(meox = ifelse(class == "meox1", "meox1", "other"))

# Generate dotplot
p1 <- DotPlot_scCustom(micdrop, 
                       idents = som_ct, 
                       features = meox_genes, 
                       scale = FALSE, 
                       group.by = 'meox', 
                       flip_axes = TRUE) +
  # Add black borders to points
  geom_point(aes(size = pct.exp), shape = 21, colour = "black", stroke = 0.5) +
  # Apply magma color scheme
  scale_colour_viridis(option = "magma", direction = -1) +
  # Customize legend
  guides(size = guide_legend(override.aes = list(shape = 21, 
                                                   colour = "black", 
                                                   fill = "white"))) + 
  # Format y-axis labels with italic for meox1
  scale_y_discrete(labels = c(expression(italic(meox1), "other"))) +
  # Theme customization
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10, 
                                    family = "Arial"),
        axis.text.y = element_text(size = 10, family = "Arial", 
                                    face = "italic"),                        
        legend.title = element_text(size = 10, family = "Arial"),                       
        legend.text = element_text(size = 10, family = "Arial"),                        
        axis.title = element_text(size = 10, family = "Arial")) + 
  ylab(element_blank()) + 
  xlab("Mutant")

p1

# Save figure
ggsave(filename = "outputs_fig3/Figure3A.svg", 
       device = "svg",
       width = 3, 
       height = 3,
       units = 'in',
       dpi = "retina",
       bg = "transparent")
```

# ============================================================================
# SECTION 2: FEATURE PLOTS - GENE EXPRESSION ON UMAP
# ============================================================================

## Figure 3B: meox1 and comp expression
```{r figure3B-featureplots}
# Create feature plots for meox1 and comp
p1 <- FeaturePlot(micdrop, features = "meox1", reduction = "umap") + 
  NoAxes() + 
  theme(plot.title = element_text(size = 12, face = "italic"))

p2 <- FeaturePlot(micdrop, features = "comp", reduction = "umap") + 
  NoAxes() + 
  theme(plot.title = element_text(size = 12, face = "italic"))

# Define shared color scale (grey to magma gradient)
sc <- scale_color_gradientn(
  colors = c("lightgrey", viridis(100, option = "magma", direction = -1)), 
  limits = c(0, 4)
) 

# Apply consistent formatting
p3 <- p1 + sc + 
  theme(plot.title = element_text(size = 12, face = "italic"),
        legend.position = "bottom",
        legend.direction = "horizontal")

p4 <- p2 + sc + 
  theme(plot.title = element_text(size = 12, face = "italic"),
        legend.position = "bottom",
        legend.direction = "horizontal")

# Save meox1 expression
p3 + NoLegend() + ggtitle("")
ggsave(filename = "outputs_fig3/Figure3B.png", 
       device = "png",
       width = 3, 
       height = 3,
       units = 'in',
       dpi = "retina",
       bg = "transparent")

# Save comp expression
p4 + NoLegend() + ggtitle("")
ggsave(filename = "outputs_fig3/Figure3B-2.png", 
       device = "png",
       width = 3, 
       height = 3,
       units = 'in',
       dpi = "retina",
       bg = "transparent")
```

## Figure 3F: gata2a expression
```{r figure3F-gata2a}
p1 <- FeaturePlot(micdrop, features = "gata2a", reduction = "umap") + 
  NoAxes() + 
  theme(plot.title = element_text(size = 12, face = "italic"))

p1 + sc + 
  theme(plot.title = element_text(size = 12, face = "italic"),
        legend.position = "bottom",
        legend.direction = "horizontal") 

ggsave(filename = "outputs_fig3/Figure3F.svg", 
       device = "svg",
       width = 3, 
       height = 4,
       units = 'in',
       dpi = "retina",
       bg = "transparent")
```

## Figure 3J: dlx1a expression
```{r figure3J-dlx1a}
p1 <- FeaturePlot(micdrop, features = "dlx1a", reduction = "umap") + 
  NoAxes() + 
  theme(plot.title = element_text(size = 12, face = "italic"))

p1 + sc + 
  theme(plot.title = element_text(size = 12, face = "italic"),
        legend.position = "bottom",
        legend.direction = "horizontal") 

ggsave(filename = "outputs_fig3/Figure3J.svg", 
       device = "svg",
       width = 3, 
       height = 4,
       units = 'in',
       dpi = "retina",
       bg = "transparent")
```

# ============================================================================
# SECTION 3: CELL COMPOSITION ANALYSIS - BOXPLOTS
# ============================================================================

## Figure 3E: Optic Tectum cell percentages
```{r figure3E-tectum-boxplot}
# 1. Prepare the dataframe first

meta <- micdrop@meta.data


counts <- meta %>% 
  group_by(class, group) %>% 
  dplyr::count(curated_cell_type2) %>%
  mutate(total_perturb = sum(n),
         pct_cluster = n / total_perturb ) 

plot_data <- counts %>% 
  filter(curated_cell_type2 == "Midbrain - Optic Tectum") %>%
  mutate(class = ifelse(class %in% c("sox11b", "mafba", "neurod1", "nkx2.7"), 
                        class, "other")) %>%
  mutate(class = factor(class, levels = c("other", "sox11b", "mafba", 
                                          "neurod1", "nkx2.7")))

# 2. Plot
ggplot(plot_data, aes(x = class, y = pct_cluster * 100)) +
  # Boxplot uses ALL data (plot_data)
  geom_boxplot(fill = 'grey', outliers = FALSE, size = 0.2) + 
  
  # Jitter uses ONLY data where class is NOT 'other'
  geom_jitter(data = subset(plot_data, class != "other"), 
              width = 0.2,  
              size = 0.35) + 
  
  ylim(c(0, 2.0)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10),
        axis.text.y = element_text(size = 10, family = "Arial"),
        axis.title.x = element_text(size = 10, family = "Arial"),
        axis.title.y = element_text(size = 10, family = "Arial"),
        axis.line = element_line(size = 0.2)) +
  labs(y = "% cells in cluster", x = "Mutation") +
  scale_x_discrete(labels = c("other" = "other", 
                              "sox11b" = expression(italic("sox11b")),
                              "mafba" = expression(italic("mafba")),
                              "neurod1" = expression(italic("neurod1")),
                              "nkx2.7" = expression(italic("nkx2.7"))))

ggsave(filename = "outputs_fig3/Figure3E.eps", 
       device = "eps",
       width = 1.75, 
       height = 2.5,
       units = 'in',
       bg = "transparent")
```

## Figure 3I: Hypothalamus cell percentages
```{r figure3I-hypothalamus-boxplot}
# 1. Prepare the dataframe 
plot_data_hypo <- counts %>% 
  filter(curated_cell_type2 == "Diencephalon - Hypothalamus / POA") %>%
  mutate(class = ifelse(class %in% c("sox11b", "rx3"), class, "other")) %>%
  mutate(class = factor(class, levels = c("other", "sox11b", "rx3")))

# 2. Plot
ggplot(plot_data_hypo, aes(x = class, y = pct_cluster * 100)) +
  # Boxplot uses ALL data
  geom_boxplot(fill = 'grey', outliers = FALSE, size = 0.2) + 
  
  # Jitter uses ONLY data where class is NOT 'other'
  geom_jitter(data = subset(plot_data_hypo, class != "other"), 
              width = 0.2, 
              size = 0.35) +
  
  ylim(c(0, 4)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10),
        axis.text.y = element_text(size = 10, family = "Arial"),
        axis.title.x = element_text(size = 10, family = "Arial"),
        axis.title.y = element_text(size = 10, family = "Arial"),
        axis.line = element_line(size = 0.2)) +
  labs(y = "% cells in cluster", x = "Mutation") +
  
  # Cleaned up labels to match only the groups present in this plot
  scale_x_discrete(labels = c("other" = "other", 
                              "sox11b" = expression(italic("sox11b")),
                              "rx3" = expression(italic("rx3"))))

ggsave(filename = "outputs_fig3/Figure3I.eps", 
       device = "eps",
       width = 1, 
       height = 2.5,
       units = 'in',
       dpi = "retina",
       bg = "transparent")
```

# ============================================================================
# SECTION 4: IN SITU HYBRIDIZATION AREA MEASUREMENTS
# ============================================================================

## Load and prepare ISH measurement data
```{r load-ish-measurements}
# Load gata2a area measurements
gata <- read.csv(file = "input_data/gata_measurements.csv") %>%
  filter(probe == "gata")

# Normalize area to head width and reorder factor levels
gata <- gata %>%
  mutate(head_size_factor = mean(head.width) / head.width) %>%
  mutate(area.normalized = stain.area * head_size_factor) %>%
  mutate(injection = factor(genotype, 
                            levels = c("scramble", "sox11b", "mafba", 
                                       "neurod1", "nkx2.7")))

# Load dlx1a area measurements  
dlx <- read.csv(file = "input_data/dlx_measurements.csv")

# Normalize area to trunk area and reorder factor levels
dlx <- dlx %>%
  mutate(head_size_factor = mean(head_width) / head_width) %>%
  mutate(area.normalized = pixel_area * head_size_factor) %>%
  mutate(injection = factor(injection, 
                            levels = c("scramble", "sox11b", "rx3")))
```

## Figure 3H: gata2a area boxplot
```{r figure3H-gata-area}
ggplot(gata, aes(x = injection, y = area.normalized)) +
  geom_boxplot(fill = "lightgrey", size = 0.2, outliers = FALSE) + 
  geom_jitter(width = 0.2, size = 0.35, color = "black") + 
  theme_classic() +
  theme(axis.text.x = element_text(size = 10, angle = 45, hjust = 1), 
        axis.text.y = element_text(size = 10), 
        axis.title.x = element_text(size = 10), 
        axis.title.y = element_text(size = 10),
        axis.line = element_line(size = 0.2)) + 
  labs(x = "Mutation", y = "Normalized Area") +
  scale_x_discrete(labels = c("scramble" = "control", 
                              "sox11b" = expression(italic("sox11b")),
                              "mafba" = expression(italic("mafba")),
                              "neurod1" = expression(italic("neurod1")),
                              "nkx2.7" = expression(italic("nkx2.7")))) +
  ylim(c(0,9000))

ggsave(filename = "outputs_fig3/Figure3H.eps", 
       device = "eps",
       width = 2, 
       height = 2.5,
       units = 'in',
       dpi = "retina",
       bg = "transparent") 
```

## Figure 3L: dlx1a area boxplot
```{r figure3L-dlx-area}
ggplot(dlx, aes(x = injection, y = area.normalized)) +
  geom_boxplot(fill = "lightgrey", size = 0.2, outliers = FALSE) + 
  geom_jitter(width = 0.1, size = 0.2, color = "black") + 
  theme_classic() +
  theme(axis.text.x = element_text(size = 10, angle = 45, hjust = 1), 
        axis.text.y = element_text(size = 10), 
        axis.title.x = element_text(size = 10), 
        axis.title.y = element_text(size = 10),
        axis.line = element_line(size = 0.2)) + 
  labs(x = "Mutation", y = "Normalized Area") +
  scale_x_discrete(labels = c("scramble" = "control", 
                              "sox11b" = expression(italic("sox11b")),
                              "rx3" = expression(italic("rx3")))) +
  ylim(0, 9000)

ggsave(filename = "outputs_fig3/Figure3L.eps", 
       device = "eps",
       width = 1.5, 
       height = 2.5,
       units = 'in',
       dpi = "retina",
       bg = "transparent") 
```

## Statistical analysis: Area measurements (ANOVA + Dunnett's test)
```{r stats-ish-area}
# gata2a area statistics
# One-way ANOVA
gata_anova <- gata %>% anova_test(area.normalized ~ injection)

# Dunnett's post-hoc test (compare all groups to scramble control)
gata_dunn <- DunnettTest(x = gata$area.normalized, 
                         g = gata$injection, 
                         control = "scramble")

# Display results
cat("Sample table gata2a - n values per sample")
print(table(gata$injection))
cat("\n=== gata2a Area ANOVA ===\n")
print(gata_anova)
cat("\n=== gata2a Area Dunnett's Test ===\n")
print(gata_dunn)

# dlx1a area statistics
# One-way ANOVA
dlx_anova <- dlx %>% anova_test(area.normalized ~ injection)

# Dunnett's post-hoc test (compare all groups to scramble control)
dlx_dunn <- DunnettTest(x = dlx$area.normalized, 
                        g = dlx$injection, 
                        control = "scramble")

# Display results
cat("Sample table dlx1a - n values per sample")
print(table(dlx$injection))
cat("\n=== dlx1a Area ANOVA ===\n")
print(dlx_anova)
cat("\n=== dlx1a Area Dunnett's Test ===\n")
print(dlx_dunn)
```

# ============================================================================
# SECTION 5: SUPPLEMENTARY FIGURES - EXPRESSION VALIDATION
# ============================================================================

## Figure S3.1: gata2a expression in Optic Tectum
```{r figureS3-1-gata-violin}
# Create categorical variable for specific mutations in Optic Tectum
micdrop@meta.data <- micdrop@meta.data %>%
  mutate(gata_tectum = ifelse(class %in% c("sox11b", "mafba", 
                                            "neurod1", "nkx2.7"), 
                               class, "other")) %>%
  mutate(gata_tectum = factor(gata_tectum, 
                               levels = c("other", "sox11b", "mafba", 
                                          "neurod1", "nkx2.7")))

# Violin plot of gata2a expression
VlnPlot(micdrop, 
        features = 'gata2a', 
        idents = "Midbrain - Optic Tectum", 
        group.by = "gata_tectum", 
        cols = rep("grey", 5)) + 
  NoLegend() +
  ggtitle("gata2a expression - Optic Tectum") +
  theme(plot.title = element_text(size = 10)) + 
  scale_x_discrete(labels = c("other" = "other", 
                              "sox11b" = expression(italic(sox11b)),
                              "mafba" = expression(italic(mafba)),
                              "neurod1" = expression(italic(neurod1)),
                              "nkx2.7" = expression(italic(nkx2.7))))

ggsave(filename = "outputs_fig3/FigureS3.1.eps", 
       device = "eps",
       width = 3, 
       height = 3,
       units = 'in',
       dpi = "retina",
       bg = "transparent")       
```

## Figure S3.1B: dlx1a expression in Hypothalamus
```{r figureS3-1B-dlx-violin}
# Create categorical variable for specific mutations in Hypothalamus
micdrop@meta.data <- micdrop@meta.data %>%
  mutate(dlx_hypo = ifelse(class %in% c("sox11b", "rx3"), class, "other")) %>%
  mutate(dlx_hypo = factor(dlx_hypo, levels = c("other", "sox11b", "rx3")))

# Violin plot of dlx1a expression
VlnPlot(micdrop, 
        features = 'dlx1a', 
        idents = "Diencephalon - Hypothalamus / POA", 
        group.by = "dlx_hypo", 
        cols = rep("grey", 3)) + 
  NoLegend() +
  ggtitle("dlx1a expression - Hypothalamus / POA") +
  theme(plot.title = element_text(size = 10)) + 
  scale_x_discrete(labels = c("other" = "other", 
                              "sox11b" = expression(italic(sox11b)),
                              "rx3" = expression(italic(rx3))))

ggsave(filename = "outputs_fig3/FigureS3.1B.eps", 
       device = "eps",
       width = 3, 
       height = 3,
       units = 'in',
       dpi = "retina",
       bg = "transparent")  
```

## Verify no significant DEG for probe genes
```{r verify-deg-results}
# Reference DEG results to ensure no significant difference between 
# mutants for dlx1a and gata2a probed genes

# Load differential expression results
DEG_results <- readRDS(file = "input_data/DEG_results_5_9_24.rds")

# Extract dlx1a test results for relevant mutations in Hypothalamus
dlx1a_test_results <- DEG_results %>% 
  filter(perturbation %in% c("rx3", "sox11b"), 
         cell_type == "Diencephalon - Hypothalamus", 
         gene == "dlx1a")

# Extract gata2a test results for relevant mutations in Optic Tectum
gata2a_test_results <- DEG_results %>% 
  filter(perturbation %in% c("mafba", "sox11b", "neurod1", "nkx2.7"), 
         cell_type == "Midbrain - Optic Tectum", 
         gene == "gata2a")

cat("\n=== dlx1a Expression Test Results ===\n")
print(dlx1a_test_results)

cat("\n=== gata2a Expression Test Results ===\n")
print(gata2a_test_results)
```

# ============================================================================
# SECTION 6: RHOMBOMERE WIDTH MEASUREMENTS
# ============================================================================

## Load and prepare width measurement data
```{r load-width-measurements}
# Load egr2b rhombomere measurements
egr2b_measure <- read.csv(file = "input_data/egr2b_measure_repeat2.csv")

# Normalize R3 and R5 widths to trunk width
egr2b_measure <- egr2b_measure %>% 
  mutate(R3_norm = R3_width / trunk_width, 
         R5_norm = R5_width / trunk_width) %>%
  # Reorder factor levels for plotting
  mutate(Mutant = factor(Mutant, 
                         levels = c("scramble", "sox11b", "rx3", 
                                    "neurod1", "mafba", "nkx2.7")))

# Load ntl (notochord) measurements
ntl_measure <- read.csv(file = "input_data/ntl_measurements.csv")

# Normalize notochord width to trunk width
ntl_measure <- ntl_measure %>% 
  mutate(ntc_norm = ntc_width / trunk_width) %>%
  # Reorder factor levels for plotting
  mutate(Mutant = factor(Mutant, 
                         levels = c("scramble", "sox11b", "rx3", 
                                    "neurod1", "mafba", "nkx2.7")))
```

## Figure S9D: Rhombomere 3 width
```{r figureS9D-R3-width}

ggplot(egr2b_measure, aes(x = Mutant, y = R3_norm)) +
  geom_boxplot(fill = "lightgrey", size = 0.2, outliers = FALSE) + 
  geom_jitter(width = 0.1, size = 0.2, color = "black") + 
  theme_classic() +
  theme(axis.text.x = element_text(size = 10, angle = 45, hjust = 1), 
        axis.text.y = element_text(size = 10), 
        axis.title.x = element_text(size = 10), 
        axis.title.y = element_text(size = 10),
        axis.line = element_line(size = 0.2)) + 
  labs(x = "Mutation", y = "Normalized Width") +
  scale_x_discrete(labels = c("scramble" = "control", 
                              "sox11b" = expression(italic("sox11b")),
                              "mafba" = expression(italic("mafba")),
                              "neurod1" = expression(italic("neurod1")),
                              "nkx2.7" = expression(italic("nkx2.7")),
                              "rx3" = expression(italic("rx3")))) +
  ylim(0, 1)

ggsave(filename = "outputs_fig3/FigureS9D.eps", 
       device = "eps",
       width = 2.5, 
       height = 2.5,
       units = 'in',
       dpi = "retina",
       bg = "transparent")  

```

## Figure S9E: Rhombomere 5 width
```{r figureS9E-R5-width}
ggplot(egr2b_measure, aes(x = Mutant, y = R5_norm)) +
  geom_boxplot(fill = "lightgrey", size = 0.2, outliers = FALSE) + 
  geom_jitter(width = 0.1, size = 0.2, color = "black") + 
  theme_classic() +
  theme(axis.text.x = element_text(size = 10, angle = 45, hjust = 1), 
        axis.text.y = element_text(size = 10), 
        axis.title.x = element_text(size = 10), 
        axis.title.y = element_text(size = 10),
        axis.line = element_line(size = 0.2)) + 
  labs(x = "Mutation", y = "Normalized Width") +
  scale_x_discrete(labels = c("scramble" = "control", 
                              "sox11b" = expression(italic("sox11b")),
                              "mafba" = expression(italic("mafba")),
                              "neurod1" = expression(italic("neurod1")),
                              "nkx2.7" = expression(italic("nkx2.7")),
                              "rx3" = expression(italic("rx3")))) +
  ylim(0, 1)

ggsave(filename = "outputs_fig3/FigureS9E.eps", 
       device = "eps",
       width = 2.5, 
       height = 2.5,
       units = 'in',
       dpi = "retina",
       bg = "transparent")  
```

## Figure S9G: Notochord width
```{r figureS9G-notochord-width}
ggplot(ntl_measure, aes(x = Mutant, y = ntc_norm)) +
  geom_boxplot(fill = "lightgrey", size = 0.2, outliers = FALSE) + 
  geom_jitter(width = 0.1, size = 0.2, color = "black") + 
  theme_classic() +
  theme(axis.text.x = element_text(size = 10, angle = 45, hjust = 1), 
        axis.text.y = element_text(size = 10), 
        axis.title.x = element_text(size = 10), 
        axis.title.y = element_text(size = 10),
        axis.line = element_line(size = 0.2)) + 
  labs(x = "Mutation", y = "Normalized Width") +
  scale_x_discrete(labels = c("scramble" = "control", 
                              "sox11b" = expression(italic("sox11b")),
                              "mafba" = expression(italic("mafba")),
                              "neurod1" = expression(italic("neurod1")),
                              "nkx2.7" = expression(italic("nkx2.7")),
                              "rx3" = expression(italic("rx3")))) +
  ylim(0, 1)

ggsave(filename = "outputs_fig3/FigureS9G.eps", 
       device = "eps",
       width = 2.5, 
       height = 2.5,
       units = 'in',
       dpi = "retina",
       bg = "transparent")  
```

## Statistical analysis: Width measurements (ANOVA + Dunnett's test)
```{r stats-width-measurements}
# Rhombomere 5 width statistics
R5_anova <- egr2b_measure %>% anova_test(R5_norm ~ Mutant)
dunn_R5 <- DunnettTest(x = egr2b_measure$R5_norm, 
                       g = egr2b_measure$Mutant, 
                       control = "scramble")

# Rhombomere 3 width statistics
R3_anova <- egr2b_measure %>% anova_test(R3_norm ~ Mutant)
dunn_R3 <- DunnettTest(x = egr2b_measure$R3_norm, 
                       g = egr2b_measure$Mutant, 
                       control = "scramble")

# Notochord width statistics
ntc_anova <- ntl_measure %>% anova_test(ntc_norm ~ Mutant)
dunn_ntc <- DunnettTest(x = ntl_measure$ntc_norm, 
                        g = ntl_measure$Mutant, 
                        control = "scramble")

# Display all results
cat("Sample table - n embryos")
print(table(egr2b_measure$Mutant))
cat("\n=== Rhombomere 5 Width ANOVA ===\n")
print(R5_anova)
cat("\n=== Rhombomere 5 Width Dunnett's Test ===\n")
print(dunn_R5)

cat("Sample table - n embryos")
print(table(egr2b_measure$Mutant))
cat("\n=== Rhombomere 3 Width ANOVA ===\n")
print(R3_anova)
cat("\n=== Rhombomere 3 Width Dunnett's Test ===\n")
print(dunn_R3)

cat("Sample table - n embryos")
print(table(ntl_measure$Mutant))
cat("\n=== Notochord Width ANOVA ===\n")
print(ntc_anova)
cat("\n=== Notochord Width Dunnett's Test ===\n")
print(dunn_ntc)
```

```{r}


# Rhombomere 5 width statistics
R5_anova <- egr2b_measure %>% anova_test(R5_norm ~ Mutant)
dunn_R5 <- DunnettTest(x = egr2b_measure$R5_norm, 
                       g = egr2b_measure$Mutant, 
                       control = "scramble")

# Rhombomere 3 width statistics
R3_anova <- egr2b_measure %>% anova_test(R3_norm ~ Mutant)
dunn_R3 <- DunnettTest(x = egr2b_measure$R3_norm, 
                       g = egr2b_measure$Mutant, 
                       control = "scramble")

# Display all results
cat("\n=== Rhombomere 5 Width ANOVA ===\n")
print(R5_anova)
cat("\n=== Rhombomere 5 Width Dunnett's Test ===\n")
print(dunn_R5)

cat("\n=== Rhombomere 3 Width ANOVA ===\n")
print(R3_anova)
cat("\n=== Rhombomere 3 Width Dunnett's Test ===\n")
print(dunn_R3)
```
```{r}
# Filter for cell types with >800 cells
celltypes_keep <- micdrop@meta.data %>%
  dplyr::count(curated_cell_type2) %>%
  filter(n > 800) %>%
  pull(curated_cell_type2)

# Calculate mean percentage per genotype per cluster (across 4 replicates)
heatmap_data <- micdrop@meta.data %>%
  filter(curated_cell_type2 %in% celltypes_keep) %>%
  group_by(class, group, curated_cell_type2) %>%
  summarize(n = n(), .groups = "drop") %>%
  group_by(class, group) %>%
  mutate(pct = n / sum(n) * 100) %>%
  ungroup() %>%
  # Average across the 4 replicates
  group_by(class, curated_cell_type2) %>%
  summarize(mean_pct = mean(pct), .groups = "drop")

# Create separate dataframes for target genotypes and "other"
target_genotypes <- heatmap_data %>%
  filter(class %in% c("sox11b", "mafba", "neurod1", "nkx2.7", "rx3")) %>%
  pivot_wider(names_from = curated_cell_type2, 
              values_from = mean_pct, 
              values_fill = 0) %>%
  column_to_rownames(var = "class")

# Calculate "other" as mean of all non-target genotypes
other_data <- heatmap_data %>%
  filter(!class %in% c("sox11b", "mafba", "neurod1", "nkx2.7", "rx3")) %>%
  group_by(curated_cell_type2) %>%
  summarize(mean_pct = mean(mean_pct), .groups = "drop") %>%
  pivot_wider(names_from = curated_cell_type2, 
              values_from = mean_pct, 
              values_fill = 0)

# Add "other" as a row
rownames(other_data) <- "curated_cell_type2"
other_row <- other_data %>%
  t() %>%
  as.data.frame()
colnames(other_row) <- "other"
other_row <- t(other_row)

# Combine with target genotypes in specified order
row_order <- c("other", "neurod1", "mafba", "nkx2.7", "sox11b", "rx3")
heatmap_matrix <- rbind(other_row, as.matrix(target_genotypes))
heatmap_matrix <- heatmap_matrix[row_order, ]

# Color scale
library(circlize)
col_fun4 <- colorRamp2(
  breaks = seq(0, 4, length.out = 100),
  colors = viridis(100, option = "mako")
)

# Find column indices
hypothalamus_col <- which(colnames(heatmap_matrix) == "Diencephalon - Hypothalamus")
if (length(hypothalamus_col) == 0) {
  hypothalamus_col <- grep("Hypothalamus", colnames(heatmap_matrix), value = FALSE)
}

tectum_col <- which(colnames(heatmap_matrix) == "Midbrain - Optic Tectum")
if (length(tectum_col) == 0) {
  tectum_col <- grep("Optic Tectum", colnames(heatmap_matrix), value = FALSE)
}

# Create cell highlighting function
cell_fun <- function(j, i, x, y, width, height, fill) {
  if (length(hypothalamus_col) > 0 && length(tectum_col) > 0) {
    # Hypothalamus: sox11b and rx3
    if (j == hypothalamus_col && i %in% match(c("sox11b", "rx3"), rownames(heatmap_matrix))) {
      grid.rect(x, y, width, height, 
                gp = gpar(col = "darkgrey", lwd = 3, fill = NA))
    }
    # Optic Tectum: sox11b, mafba, neurod1, nkx2.7
    if (j == tectum_col && i %in% match(c("sox11b", "mafba", "neurod1", "nkx2.7"), 
                                         rownames(heatmap_matrix))) {
      grid.rect(x, y, width, height, 
                gp = gpar(col = "darkgrey", lwd = 3, fill = NA))
    }
  }
}

# Create heatmap
hm <- Heatmap(
  heatmap_matrix,
  name = "% cells",
  col = col_fun4,
  cluster_rows = FALSE,  # Keep order as specified
  cluster_columns = TRUE,
  show_column_names = TRUE,
  show_row_names = TRUE,
  column_names_rot = 45,
  column_names_side = "bottom",
  column_names_gp = gpar(fontsize = 10),
  row_names_side = "right",
  row_names_gp = gpar(fontsize = 11, fontface = "italic"),  # Everything italic
  border = TRUE,
  cell_fun = cell_fun,
  heatmap_legend_param = list(
    title = "% cells in cluster",
    direction = "horizontal",
    legend_width = unit(6, "cm")
  )
)

# Save
png(filename = "outputs_fig3/FigureS3_celltype_percentages.png", 
    height = 5, width = 14, units = "in", res = 300)
draw(hm, heatmap_legend_side = "bottom", padding = unit(c(10, 30, 10, 10), "mm"))
dev.off()
```


